akka-http-core/src/main/scala/akka/http/impl/engine/client/PoolMasterActor.scala

```diff
@@ -0,0 +1,145 @@
+package akka.http.impl.engine.client
+
+import akka.Done
+import akka.actor.{ Actor, ActorLogging, ActorRef, DeadLetterSuppression, Deploy, NoSerializationVerificationNeeded, Props, Terminated }
+import akka.http.impl.engine.client.PoolInterfaceActor.PoolRequest
+import akka.http.impl.settings.HostConnectionPoolSetup
+import akka.http.scaladsl.HttpExt
+import akka.http.scaladsl.model.{ HttpRequest, HttpResponse }
+import akka.stream.Materializer
+
+import scala.concurrent.{ Future, Promise }
+
+/**
+ * Manages access to a host connection pool or rather: a sequence of pool incarnations.
+ *
+ * A host connection pool for a given [[HostConnectionPoolSetup]] is a running stream, whose outside interface is
+ * provided by its [[PoolInterfaceActor]] actor. The actor accepts [[PoolInterfaceActor.PoolRequest]] messages
+ * and completes their `responsePromise` whenever the respective response has been received (or an error occurred).
+ *
+ * The [[PoolMasterActor]] provides a layer of indirection between a [[PoolGateway]], which represents a pool,
+ * and the [[PoolInterfaceActor]] instances which are created on-demand and stopped after an idle-timeout.
+ *
+ * Several [[PoolGateway]] objects may be mapped to the same pool if they have the same [[HostConnectionPoolSetup]]
+ * and are marked as being shared. This is the case for example for gateways obtained through
+ * [[HttpExt.cachedHostConnectionPool]]. Some other gateways are not shared, such as those obtained through
+ * [[HttpExt.newHostConnectionPool]], and will have their dedicated restartable pool.
+ *
+ */
+private[http] class PoolMasterActor extends Actor with ActorLogging {
+
+  import PoolMasterActor._
+
+  private[this] var poolStatus = Map[PoolGateway, PoolInterfaceStatus]()
+  private[this] var poolInterfaces = Map[ActorRef, PoolGateway]()
+
+  /**
+   * Start a new pool interface actor, register it in our maps, and watch its death. No actor should
+   * currently exist for this pool.
+   *
+   * @param gateway the pool gateway this pool corresponds to
+   * @param fm the materializer to use for this pool
+   * @return the newly created actor ref
+   */
+  private[this] def startPoolInterfaceActor(gateway: PoolGateway)(implicit fm: Materializer): ActorRef = {
+    if (poolStatus.contains(gateway)) {
+      throw new IllegalStateException(s"pool interface actor for $gateway already exists")
```

Comment:
Yes, but I already feel that this is over-engineered: this method is only called if `poolStatus.get(gateway)` is `None`. If somebody modifies the code despite the documentation and the explicit redundant test in the method, I sure hope that this will break instead of failing silently.